### PR TITLE
working csharp toolchain in linux environment

### PR DESCRIPTION
**Pull request by ildarisaev**

### CODE REVIEW

**Comment by beyang** [comment]
can you include a `.travis.yml` file like in other srclib toolchains so the tests will automatically run in Travis CI?

File: .bin/srclib-csharp [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+BIN_DIR="`dirname \"$BASH_SOURCE\"`"
+ROOT_DIR="`dirname \"$BIN_DIR\"`"
+PROJ_DIR="$ROOT_DIR/Srclib.Nuget"
+#echo "$PROJ_DIR"
+
+#echo dnx -p "$PROJ_DIR/project.json" run $*
+dnx -p "$PROJ_DIR" run $*
```
Comment (beyang) [inline .bin/srclib-csharp:9]: Add a `Makefile` with a target `dep` that installs the necessary dependencies to run the toolchain.

File: Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        _dnuPath = new Lazy<string>(FindDnuNix);
+      }
+
+      cmdApp.Command("depresolve", c => {
+        c.Description = "Perform a combination of parsing, static analysis, semantic analysis, and type inference";
+
+        c.HelpOption("-?|-h|--help");
+
+        c.OnExecute(async () => {
+          //System.Diagnostics.Debugger.Launch();
```
Comment (beyang) [inline Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs]: delete

File: Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+      p.StartInfo.RedirectStandardOutput = true;
+      p.StartInfo.RedirectStandardError = true;
+
+      p.Start();
+      // it's important to read stdout and stderr, else it might deadlock
+      var outs = await Task.WhenAll(p.StandardOutput.ReadToEndAsync(), p.StandardError.ReadToEndAsync());
+      p.WaitForExit();
+
+      // in the future, actually parse output or something
+      //Console.WriteLine(outs[0]);
```
Comment (beyang) [inline Srclib.Nuget/ConsoleCommands/DepresolveConsoleCommand.cs]: delete

File: Srclib.Nuget/ConsoleCommands/ScanConsoleCommand.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        var repoName = c.Option ("--repo <REPOSITORY_URL>",   "The URI of the repository that contains the directory tree being scanned", CommandOptionType.SingleValue);
+        var subdir   = c.Option ("--subdir <SUBDIR_PATH>", "The path of the current directory (in which the scanner is run), relative to the root directory of the repository being scanned", CommandOptionType.SingleValue);
+
+        c.HelpOption("-?|-h|--help");
+
+        c.OnExecute(() => {
+          var repository = repoName.Value();
+          var dir = Path.GetFullPath(Path.Combine(Directory.GetCurrentDirectory(), subdir.Value()));
+
+          //Console.WriteLine($"Repository: {repository}");
```
Comment (beyang) [inline Srclib.Nuget/ConsoleCommands/ScanConsoleCommand.cs]: delete these commented-out lines

File: Srclib.Nuget/ConsoleCommands/ScanConsoleCommand.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+          //Console.WriteLine($"Directory: {dir}");
+
+          var sourceUnits = new List<SourceUnit>();
+          foreach(var proj in Scan(dir))
+          {
+            //Console.WriteLine($"Found project: {proj.Name} ({proj.Version}, {proj.CompilerServices})");
```
Comment (beyang) [inline Srclib.Nuget/ConsoleCommands/ScanConsoleCommand.cs]: delete

File: Srclib.Nuget/ConsoleCommands/ScanConsoleCommand.cs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+            sourceUnits.Add(SourceUnit.FromProject(proj, dir));
+          }
+
+          //Console.Error.Write($"Current dir: {Environment.CurrentDirectory}\n");
```
Comment (beyang) [inline Srclib.Nuget/ConsoleCommands/ScanConsoleCommand.cs]: delete

File: Srclib.Nuget/ConsoleCommands/ScanConsoleCommand.cs [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+          Console.WriteLine(JsonConvert.SerializeObject(sourceUnits, Formatting.Indented));
+
+          return 0;
+        });
+      });
+    }
+
+    static IEnumerable<Project> Scan(string dir)
+    {
+      //Console.WriteLine($"Scan dir: {dir}\n");
```
Comment (beyang) [inline Srclib.Nuget/ConsoleCommands/ScanConsoleCommand.cs]: delete

File: Srclib.Nuget/ConsoleCommands/ScanConsoleCommand.cs [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+      if (Project.HasProjectFile(dir))
+      {
+        //Console.WriteLine($"Dir has project: {dir}\n");
```
Comment (beyang) [inline Srclib.Nuget/ConsoleCommands/ScanConsoleCommand.cs]: delete

File: Srclib.Nuget/Documentation/DocProcessor.cs [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+
+            case XmlNodeType.Element:
+              switch (reader.LocalName)
+              {
+                case "see":
+                  written = true;
+                  var cref = reader.GetAttribute("cref");
+                  writer.WriteStartElement("span");
+
+                  // todo, lookup actual symbol.
```
Comment (beyang) [inline Srclib.Nuget/Documentation/DocProcessor.cs]: use the format `// TODO($username)`

File: Srclib.Nuget/Documentation/DocProcessor.cs [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+        if (!notEmpty)
+        {
+          return null;
+        }
+      }
+
+      return sb.ToString();
+    }
+
+    static bool ProcessContent(XElement node, XmlWriter writer)
```
Comment (beyang) [inline Srclib.Nuget/Documentation/DocProcessor.cs:174]: Can you add a docstring to every exported function/method/class?

File: Srclib.Nuget/Graph/GraphContext.cs [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+{
+  public class GraphContext
+  {
+    public SourceUnit SourceUnit { get; set; }
+    public string ProjectDirectory { get; set; }
+    public Project Project { get; set; }
+    public ApplicationHostContext ApplicationHostContext { get; set; }
+    public Microsoft.Extensions.PlatformAbstractions.IApplicationEnvironment HostEnvironment { get; set; }
+    public Microsoft.Extensions.PlatformAbstractions.IApplicationEnvironment ApplicationEnvironment { get; set; }
+    //public IFileWatcher FileWatcher { get; set; }
```
Comment (beyang) [inline Srclib.Nuget/Graph/GraphContext.cs]: delete

File: Srclib.Nuget/Graph/GraphRunner.cs [hunk 12]
```diff
@@ -0,0 +1,446 @@
...
+        TargetFramework = context.Project.GetTargetFrameworks().First().FrameworkName
+      };
+
+      context.ApplicationEnvironment = new ApplicationEnvironment(
+        context.Project,
+        context.ApplicationHostContext.TargetFramework,
+        "Debug",
+        context.HostEnvironment);
+
+            //Microsoft.Extensions.PlatformAbstractions.IApplicationEnvironment f = null;
```
Comment (beyang) [inline Srclib.Nuget/Graph/GraphRunner.cs]: delete

File: .bin/srclib-csharp (hunk 1 above)
Comment (ildarisaev) [inline .bin/srclib-csharp:9]: I added a function to srclib/cli/toolchain_cmd.go that installs csharp toolchain and some of the dependencies (it's not committed anywhere yet). Isn't it a canonical way to deal with it?

File: Srclib.Nuget/Program.cs [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    public int Main(string[] args)
+    {
+      var app = new CommandLineApplication(throwOnUnexpectedArg: true);
+      app.Name = "srclib-csharp";
+      app.FullName = "Scrlib C# toolchain";
+
+      app.HelpOption("-?|-h|--help");
+      //app.VersionOption("--version", () => _runtimeEnv.GetShortVersion(), () => _runtimeEnv.GetFullVersion());
```
Comment (beyang) [inline Srclib.Nuget/Program.cs]: delete

File: Srclib.Nuget/Utils.cs [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+      {
+        folder += Path.DirectorySeparatorChar;
+      }
+      Uri folderUri = new Uri(folder);
+      String s = Uri.UnescapeDataString(folderUri.MakeRelativeUri(pathUri).ToString().Replace('/', Path.DirectorySeparatorChar));
+      return s;
+    }
+  }
+
+  // TODO: Remove, for testing purposes only
```
Comment (beyang) [inline Srclib.Nuget/Utils.cs]: remove?

**Comment by beyang** [comment]
LGTM other than comments.

File: .bin/srclib-csharp (hunk 1 above)
Comment (beyang) [inline .bin/srclib-csharp:9]: For ease of development, it's nice to have those commands in a Makefile. I think most of the other toolchains do that. Then, the function in `srclib/cli/toolchain_cmd.go` invokes `make dep`.

Comment (ildarisaev) [inline .bin/srclib-csharp:9]: I'm still not sure this is a convenient way. As far as I can see, none of the toolchains has dep target in their Makefiles

https://github.com/sourcegraph/srclib-java/blob/master/Makefile

https://github.com/sourcegraph/srclib-python/blob/master/Makefile

https://github.com/sourcegraph/srclib-go/blob/master/Makefile

https://github.com/sourcegraph/srclib-basic/blob/master/Makefile

Java toolchain doesn't try to install jdk and jre, go toolchain doesn't try to install go compiler and so on. Similarly, csharp toolchain shouldn't try to set up dnx environment. If we add installation to Makefile, then we will have to invoke make with super user privileges:

sudo make dep

If srclib/cli/toolchain_cmd.go invokes 'make dep', then we will have to run srclib with super user privileges, like:

sudo srclib toolchain install csharp

Comment (ildarisaev) [inline .bin/srclib-csharp:9]: I added dep target that loads local project dependencies (mainly standard runtime libraries).

Comment (beyang) [inline .bin/srclib-csharp:9]: Sorry, I meant an `install` target. Can you add that?

Can you also add instructions for installing dependencies to set up the development environment in the `README`?

Comment (ildarisaev) [inline .bin/srclib-csharp:9]: Surprisingly, with dnx there is no explicit build or installation. Building is done implicitly during first run or after modifications in source files. Weird that is, but from user's point of view install = run.

Comment (beyang) [inline .bin/srclib-csharp:9]: how does one install `dnx`?

Comment (ildarisaev) [inline .bin/srclib-csharp:9]: This is fairly easy and described here: http://docs.asp.net/en/latest/getting-started/installing-on-linux.html

I copied dnx installation instructions into README.md

**Comment by beyang** [comment]
Great, thanks for addressing the comments. As a final change, could you squash all the "travis experimenting" commits into one commit and then merge?

**Pull request merged**